### PR TITLE
Add versioning to coord maps and functions of time, add docs

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

The domain will be serialized and written to output files so it can be used for interpolations, as are `domain::FunctionOfTime` classes. To be able to read in domains written by older versions of the code the `pup` functions now support lightweight versioning (some already did). A `version` integer is written alongside the data when serializing, and read back in when deserializing.

### Upgrade instructions

This change breaks deserialization of domains written into volume data files by older versions of the code (and is intended to prevent such breaking changes in the future). For tools that need the deserialized domain (e.g `interpolate-to-coords` or `plot-power-monitors`) use an older version of the code, or regenerate your volume data files with the upgraded code.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by nilsvu** [comment]
Clang-tidy failure is the unrelated TOMS748 issue

File: src/Domain/Domain.hpp [hunk 1]
```diff
@@ -40,9 +40,61 @@ class CoordinateMapBase;
...
+ *   } else {
+ *     p | some_data_;
+ *   }
+ *   // ...
+ * }
+ * ```
+ *
+ * Make sure that all data types you serialize in the `pup` function also
+ * support versioning. Also make sure not to keep all factory-creatable classes
+ * registered that were written by old versions of the code.
```
Comment (wthrowe) [inline src/Domain/Domain.hpp:97]: Should this sentence be the opposite?

**Review by wthrowe — CHANGES_REQUESTED** [review]
I'm surprised that nothing gives a warning for checking that an unsigned integer is non-negative, but since nothing does I'm fine with the checks.

File: src/Domain/Domain.hpp (hunk 1 above)
Comment (nilsvu) [reply src/Domain/Domain.hpp:97]: Yep! Good catch.

**Review by wthrowe — APPROVED** [review]

**Pull request merged**